### ISSUE TITLE
mxStandardizeLISRELpaths

### ISSUE THREAD

**Issue by tbates** [OpenMx/OpenMx#400]
We have `mxStandardizeRAMpaths` for RAM  models.
It would be great to add an `mxStandardizeLISRELpaths` function to support `type = "LISREL"` `mxModels`.

**Comment by mhunter1** [comment]
This is a fine idea, but I have no plans to do it in the next 1 to 10 years.  There are about 300 lines in R/MxSummary.R for `mxStandardizedRAMPaths`.  I welcome anyone interested in this feature to write it.

**Comment by RMKirkpatrick** [comment]
I actually agree with @tbates , though I'm not in any hurry to implement this function.  I do think it should be done, though.

**Comment by mcneale** [comment]
The algebras for what I think was a LISREL option SS for standardized solution are known, so perhaps it wouldn't be so difficult to implement.  A student might be able to tackle it.

**Comment by mcneale** [comment]
Yeah, the big A and S matrices are not the most efficient, in general, even if (I-A) can be inverted efficiently.  However, I don't think the LISREL formulation would be better for large models, particularly in Behavior Genetics (where most models get at least doubled in size).  Indeed, the motivation for developing classic Mx, its combination of a matrix algebra interpreter and a numerical optimizer was largely motivated by the painful struggle to convert things like A+C+E into the one and only general LISREL formula.  In the end, I think it most intuitive to use A and S matrices, which are simple and general, or to specify the model in terms of matrices that directly implement the way we think about the model.

Our addition of algebraic derivatives for the RAM-type models is operational, but as yet has not been yielding very much improvement in speed in smallish models.  Larger models might show a greater benefit.

**Comment by tbates** [comment]
> algebraic derivatives for RAM-type models is operational but as yet has not been yielding very much improvement in speed in smallish models. Larger models might show a greater benefit.

@lf-araujo might be able to test that: compare OpenMx pre and post algebraic derivatives. What OpenMx version did algebraic RAM  come in on? I can't see any mention in the git notes: is this feature on a branch somewhere?

PS: Did anyone have another  guess then about what lavaan does for speed in its implementation?

**Comment by RMKirkpatrick** [comment]
> What OpenMx version did algebraic RAM come in on? I can't see any mention in the git notes: is this feature on a branch somewhere?

Yes, it is all in my branch, 'analytDerivs'.

> @lf-araujo might be able to test that: compare OpenMx pre and post algebraic derivatives.

That's already been tested.  Automated analytic derivatives have been a huge disappointment...so much so that they will be switched OFF by default when 'analytDerivs' is merged into 'master'.

**Comment by mcneale** [comment]
There was some improvement with larger models, iirc.  Since large RAM models are being developed for multivariate analyses, I haven't given up hope on it delivering a performance benefit.

**Comment by RMKirkpatrick** [comment]
> There was some improvement with larger models, iirc.

Yes--there was only one script from `make nightly`, models/nightly/startsTestMissing.R, for which analytic derivatives improved running time by more than 5 seconds _and_ by more than 10%, and that was only with two of the three main optimizers.

That's why I'm saying the new analytic RAM derivatives should be off by default.

**Comment by mcneale** [comment]
In an ideal world, we would by default switch them on when they are expected to help, and off when they are not.  It might not be much more than the number of variables in the RAM model.

**Comment by mhunter1** [comment]
Although interesting and valuable, this conversation is pretty far afield from `mxStandardizeLISRELpaths`.

**Comment by RMKirkpatrick** [comment]
> Although interesting and valuable, this conversation is pretty far afield from mxStandardizeLISRELpaths.

I created #402 to carry on this off-topic conversation.

**Comment by RMKirkpatrick** [comment]
I do not want this issue to remain closed as "not planned", because I DO indeed think that `mxStandardizeLISRELpaths()`-like functionality ought to be implemented.  It is not high-priority, but it ought to happen _eventually_.

**Comment by mcneale** [comment]
I think it would be good to add the math here, just for the record, and so that a user could implement it easily as an add-on function.

**Comment by tbates** [comment]
## Math for LISREL Path Standardization

LISREL separates variables into endogenous manifest ($Y$), exogenous manifest ($X$), endogenous latent ($\eta$), and exogenous latent ($\xi$).

The extended LISREL model is defined by:
$$\eta = \alpha + B \eta + \Gamma \xi + \zeta$$
$$y = \tau_y + \Lambda_y \eta + \epsilon$$
$$x = \tau_x + \Lambda_x \xi + \delta$$


Let the covariance matrix of latents

$$L = \begin{pmatrix} \eta & \xi \end{pmatrix}$$ 

be: 

$$\Sigma_L = \begin{pmatrix} \Sigma_\eta & \Sigma_{\eta\xi} \\ \Sigma_{\xi\eta} & \Sigma_\xi \end{pmatrix}$$

where:
*   $\Sigma_\xi = \Phi$
*   $\Sigma_{\eta\xi} = (I - B)^{-1} \Gamma \Phi$
*   $\Sigma_\eta = (I - B)^{-1} (\Gamma \Phi \Gamma^T + \Psi) (I - B)^{-T}$

Let the covariance matrix of manifests $M = \begin{pmatrix} Y \\ X \end{pmatrix}$ be:
$$\Sigma_M = \begin{pmatrix} \Sigma_Y & \Sigma_{YX} \\ \Sigma_{XY} & \Sigma_X \end{pmatrix}$$
where:
*   $\Sigma_Y = \Lambda_y \Sigma_\eta \Lambda_y^T + \Theta_\epsilon$
*   $\Sigma_X = \Lambda_x \Sigma_\xi \Lambda_x^T + \Theta_\delta$
*   $\Sigma_{XY} = \Lambda_x \Sigma_{\xi\eta} \Lambda_y^T + \Theta_{xy}$

The standard deviations for each set of variables are:
*   $SD_Y = \sqrt{\text{diag}(\Sigma_Y)}$
*   $SD_X = \sqrt{\text{diag}(\Sigma_X)}$
*   $SD_\eta = \sqrt{\text{diag}(\Sigma_\eta)}$
*   $SD_\xi = \sqrt{\text{diag}(\Sigma_\xi)}$

Let $V_Y, V_X, V_\eta, V_\xi$ be the diagonal matrices of these standard deviations. The standardized parameters are computed as:

1.  **Factor Loadings**:
    *   $\Lambda_{y, std} = V_Y^{-1} \Lambda_y V_\eta$ (`LY`)
    *   $\Lambda_{x, std} = V_X^{-1} \Lambda_x V_\xi$ (`LX`)
2.  **Structural Coefficients**:
    *   $B_{std} = V_\eta^{-1} B V_\eta$ (`BE`)
    *   $\Gamma_{std} = V_\eta^{-1} \Gamma V_\xi$ (`GA`)
3.  **Covariances**:
    *   $\Phi_{std} = V_\xi^{-1} \Phi V_\xi^{-1}$ (`PH`)
    *   $\Psi_{std} = V_\eta^{-1} \Psi V_\eta^{-1}$ (`PS`)
    *   $\Theta_{\epsilon, std} = V_Y^{-1} \Theta_\epsilon V_Y^{-1}$ (`TE`)
    *   $\Theta_{\delta, std} = V_X^{-1} \Theta_\delta V_X^{-1}$ (`TD`)
    *   $\Theta_{xy, std} = V_X^{-1} \Theta_{xy} V_Y^{-1}$ (`TH`)
4.  **Means**:
    *   $TX_{std} = TX V_X^{-1}$ (`TX`)
    *   $TY_{std} = TY V_Y^{-1}$ (`TY`)
    *   $KA_{std} = KA V_\xi^{-1}$ (`KA`)
    *   $AL_{std} = AL V_\eta^{-1}$ (`AL`)

**Comment by tbates** [comment]
I added `mzStandardizeLISRELpaths`.

I also added helpers to `umx` ( `umxLISREL()` and plot() ) to make it easier to build LISREL models.

But in the process I think I found a bug in the backend code for `mxExpectationLISREL`: It expects equal numbers of endogenous and exogenous manifest variables, or it throws an error. New issue here: #423
